_posts/2021-12-13-kafka-segment-retention.md

```diff
@@ -0,0 +1,208 @@
+---
+layout: post
+title: "Deep dive into Apache Kafka storage internals: segments, rolling and retention"
+date: 2021-12-13
+author: paolo_patierno
+---
+
+Everyone who uses Apache Kafka knows that it behaves as a commit-log when it comes to deal with storing records.
+All records are appended at the end of each log one after the other.
+Each log is also split in segments in order to deal with older records deletion, achieving better performances and much more.
+So, as the log is a logical sequence of records, it is comprised of segments (files) which each store a subsequence of the records.
+The broker configuration allows to tweak a lot of parameters related to logs like the ones about rolling segments, records retention and so on.
+
+Maybe not everyone knows how all these parameters have an impact on the broker behaviour and the expectations that we can have in terms of storing records for longer time and making them available to consumers.
+In this blog post, we will dig more into details about how the log segments configuration and records retention work, in order to help making better decisions depending on every single use-case.
+
+<!--more-->
+
+## The topic partition structure on the disk
+
+An Apache Kafka topic is made up by partitions where the records are appended to.
+Each partition can be defined the unit of work because it is used by clients to exchange records but cannot be defined as the unit of storage.
+A partition is further split into segments which are the actual files on the disk.
+This splitting is really important for having better performance when it comes to delete records on disk or when a consumer starts to consume from a specific offset: in both cases, dealing with a single big file is slow and error prone.
+
+Looking at the broker disk, each topic partition is a directory containing the corresponding segment files and some others.
+Using the Strimzi Canary component with its producer and consumer as an example, following how a directory looks like.
+
+
+```shell
+├── __strimzi_canary-0
+│   ├── 00000000000000000000.index
+│   ├── 00000000000000000000.log
+│   ├── 00000000000000000000.timeindex
+│   ├── 00000000000000000109.index
+│   ├── 00000000000000000109.log
+│   ├── 00000000000000000109.snapshot
+│   ├── 00000000000000000109.timeindex
+```
+
+Above, the partition 0 of the topic `__strimzi_canary` used by the Strimzi Canary component; the folder contains the following files:
+
+* The `.log` file is an actual segment containing records up to a specific offset. The name of the file defines the starting offset of the records in that log.
+* The `.index` file contains an index that maps logical offset (in effect the record's id) to the byte offset of the record within the `.log` file. It is used for accessing records at specified offsets in the log without having to scan the whole `.log` file.
+* The `.timeindex` file is another index used for accessing records by timestamp in the log.
```

Comment:
What about the `.snapshot`? You have it both above and below for one of the segments. But you do not seem to explain it.